keps/sig-storage/2857-runtime-assisted-pv-mounts/README.md

```diff
@@ -0,0 +1,1767 @@
+system operations to a container runtime) to fail a NodePublishVolume call if
+the volume is already found to be published on the same node to a different
+target path and the file system cannot handle multiple simultaneous mounts.
+- Always set ReadWriteOncePod access mode for PVCs whose associated storage
+class or CSI plugin specifies deferred mounting of file system. This can be
+enforced by the mutating webhook as a check before it sets `FSDeferralStrategy`
+or spearately enforced as a policy by a dedicated validation webhook.
+- Introduce a scheduler framework plugin that implements logic similar to the
+in-tree `VolumeRestrictions` plugin that enforces ReadWriteOncePod access for a
+PVC if deferral of file system operations is in found to be in effect based on
+annotations of the associated StorageClass or CSIDriver objects.
+
+#### Handling of NodePublish secrets
+
+Mounting and management operations on certain file systems (e.g. SMB) may
+require access to secrets. If the file system deferral strategy requires a CSI
+plugin to persist parameters for mount and management operations on files at a
+known location on disk, then secrets should not be written out in an unencrypted
+form. It is therefore recommended that storage classes that back volumes whose
+mounting and management operations require secret do not specify capabilities to
+defer file system operations to a container runtime handler where the deferral
+strategy involves specifying secrets on a metadata file on disk.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### FSDeferralStrategy field in Pod Spec
+
+A new field, `FSDeferralStrategy` will be introduced in the
+`PersistentVolumeClaimVolumeSource`, `EphemeralVolumeSource` and
+`CSIVolumeSource` structs to specify whether and how filesystem mount and
+management operations for a volume can be deferred by a CSI plugin (associated
+with the volume referred by the pod) to a container runtime handler (associated
+with the pod).
+
+Conceptually, this is similar to the existing `ReadOnly` field in
+`PersistentVolumeClaimVolumeSource` in that it applies in the context of a
+specific pod referring to a specific PVC.
+
+Specifications of FSDeferralStrategy should be lowercase alphabets and numbers
+only. There are no canonical identifiers for `FSDeferralStrategy` and in this
+respect, it is similar to `FSType` fields in various existing structs.
+
+A CSI plugin does not explicitly indicate support for a specific deferral
+strategy through CSI capabilities. However, a CSI plugin is expected to return
+an error if the specified strategy is not something it can support (as is the
+case if a specified file system type on a storage class is not supported).
```

Comment:
How will kubelet react to such error? Will it fall back to the usual CSI mount?

Reply:
Mismatch of a deferral strategy goes away with the boolean field in the latest design.